libbpf-tools/biosnoop.c

```diff
@@ -0,0 +1,273 @@
+		}
+		errno = 0;
+		env.duration = strtoll(arg, NULL, 10);
+		if (errno || env.duration <= 0) {
+			fprintf(stderr, "invalid delay (in us): %s\n", arg);
+			argp_usage(state);
+		}
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void blk_fill_rwbs(char *rwbs, unsigned int op)
+{
+	int i = 0;
+
+	if (op & REQ_PREFLUSH)
+		rwbs[i++] = 'F';
+
+	switch (op & REQ_OP_MASK) {
+	case REQ_OP_WRITE:
+	case REQ_OP_WRITE_SAME:
+		rwbs[i++] = 'W';
+		break;
+	case REQ_OP_DISCARD:
+		rwbs[i++] = 'D';
+		break;
+	case REQ_OP_SECURE_ERASE:
+		rwbs[i++] = 'D';
+		rwbs[i++] = 'E';
+		break;
+	case REQ_OP_FLUSH:
+		rwbs[i++] = 'F';
+		break;
+	case REQ_OP_READ:
+		rwbs[i++] = 'R';
+		break;
+	default:
+		rwbs[i++] = 'N';
+	}
+
+	if (op & REQ_FUA)
+		rwbs[i++] = 'F';
+	if (op & REQ_RAHEAD)
+		rwbs[i++] = 'A';
+	if (op & REQ_SYNC)
+		rwbs[i++] = 'S';
+	if (op & REQ_META)
+		rwbs[i++] = 'M';
+
+	rwbs[i] = '\0';
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	const struct event *e = data;
+	char rwbs[RWBS_LEN];
+
+	if (!start_ts)
+		start_ts = e->ts;
+	blk_fill_rwbs(rwbs, e->cmd_flags);
+	printf("%-11.6f %-14.14s %-6d %-7s %-4s %-10lld %-7d ",
+		((double)e->ts - start_ts) / 1000000000,
+		e->comm, e->pid, e->disk, rwbs, e->sector, e->len);
+	if (env.queued)
+		printf("%7.3f ", e->qdelta != -1 ?
+			(double)e->qdelta / 1000000 : -1);
+	printf("%7.3f\n", (double)e->delta / 1000000);
+
+}
+
+void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	fprintf(stderr, "lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct perf_buffer_opts pb_opts;
+	struct perf_buffer *pb = NULL;
+	struct biosnoop_bpf *obj;
+	__u64 time_end = 0;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = biosnoop_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF ojbect\n");
+		return 1;
+	}
+
+	/* initialize global data (filtering options) */
+	if (env.disk)
+		strncpy((char*)obj->rodata->targ_disk, env.disk, env.disk_len);
+	obj->rodata->targ_queued = env.queued;
+
+	err = biosnoop_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+
+	obj->links.tp_btf__blk_account_io_start =
+		bpf_program__attach(obj->progs.tp_btf__blk_account_io_start);
+	err = libbpf_get_error(obj->links.tp_btf__blk_account_io_start);
+	if (err) {
+		fprintf(stderr, "failed to attach: %s\n", strerror(err));
```

Comment:
these error messages could mention which of BPF programs failed to attach, to make it easier in case something actually don't work

Reply:
Fixed, thanks